elytron/WFLY-16793-oidc-identity-propagation.adoc

```diff
@@ -0,0 +1,166 @@
+:author:            Farah Juma
+:email:             fjuma@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+For a deployment secured with OpenID Connect using the `elytron-oidc-client` subsystem,
+the subsystem automatically creates and makes use of a virtual security domain. With
+virtual security domains, it's currently possible to propagate a security identity from the
+web layer to an EJB within the same WAR. However, there are some limitations. In particular,
+the following two things are currently missing:
+
+* Ability to propagate an identity from a WAR to an EJB within a *single* EAR
+* Ability to propagate an identity from a WAR within an EAR to an EJB within *another* EAR
+(i.e., across deployments)
+
+This RFE is to add support for these two things.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-16793[WFLY-16793]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1971[EAP7-1971]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+
+TBD
+
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* WildFly Core
+* WildFly
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* It will be possible to propagate an identity from a WAR to an EJB within a single EAR without
+requiring any new configuration to be added to enable this behaviour. This will just automatically
+work.
+
+* A new `virtual-security-domain` resource will be introduced in the `elytron` subsystem.
+The new resource will make it possible to outflow security identities established by virtual
+security domains to other security domains. The new `virtual-security-domain` resource will have
+two attributes:
+
+** `name` - This is the runtime name of a deployment associated with a virtual security domain (e.g.,
+`DEPLOYMENT_NAME.ear`, a deployment that has a subdeployment that is secured using OIDC).
+
+** `outflow-security-domains` - This is the list of `security-domains` that security identities from
+the virtual security domain should be automatically outflowed to. This will be a required attribute
+and must be non-empty.
+
+** As an example, a `virtual-security-domain` could be added as follows:
+[source]
+----
+/subsystem=elytron/virtual-security-domain=DEPLOYMENT_NAME.ear:add(outflow-security-domains=...)
+----
+
+* A new `trusted-virtual-security-domains` attribute will be added to the `security-domain` resource
```

Comment:
Will the new `virtual-security-domain` resource provide the `org.wildfly.security.security-domain` capability or a new one?

If the existing one, can we leverage `trusted-security-domains` instead of add the new `trusted-virtual-security-domains` attribute?

Reply:
Currently, when a virtual security domain is created, the `SecurityDomain` service that is provided has a name like `DEPLOYMENT_NAME.ear.security-domain.virtual`. So although it provides a `SecurityDomain`, the service name is formatted differently from a normal `security-domain` that's been configured.

We could update this so that the service name for the virtual case is of the usual form `org.wildfly.security.security-domain.DEPLOYMENT_NAME.ear`. That way, a separate `trusted-virtual-security-domains` attribute would not be needed. Instead, the existing `trusted-security-domains` attribute could be used since we'd be able to determine the service name the usual way. WDYT?

Reply:
I'm not familiar with such internals and possible impacts of such change, but it sounds reasonable and makes sense to me. It would definitively help us eliminate one extra attribute and keep the configuration relatively simple.

Reply:
+1. I've updated the analysis to mention this.

Reply:
Looked into this some more and I think we'll need the `trusted-virtual-security-domains` attribute after all. Once a `virtual-security-domain` has been added to the elytron subsystem, the actual `SecurityDomain` service that corresponds to this won't be created until the deployment unit processors get run for the deployment. This means that attempting to reference the corresponding security domain from the `trusted-security-domains` attribute won't be possible because that service won't actually be available yet. Instead, we can reference the `virtual-security-domain` from a new `trusted-virtual-security-domains` attribute. Thus, the complete configuration would look something like:

```
/subsystem=elytron/virtual-security-domain=myapp.ear:add(outflow-security-domains=[MyDomain])
/subsystem=elytron/security-domain=MyDomain:write-attribute(name=trusted-virtual-security-domains,value=[myapp.ear])
```

Reply:
OK, thanks for the clarification.  Do I understand it well that something like

```/subsystem=elytron/virtual-security-domain=myapp.ear:add(outflow-security-domains=[MyDomain])```

will work even without `myapp.ear` being deployed?

Would it be useful (and possible) to check whether all security domains set in `outflow-security-domains` have also corresponding `trusted-virtual-security-domains` set and vice versa? It could help user to identify wrong configuration for example by logging such findings.

Reply:
Yes, the `virtual-security-domain` resource can be added without the application being deployed (similar to the `secure-deployment` resource in the `elytron-oidc-client` subsystem).

Adding the checks you mentioned is an interesting idea but we don't really have a way to perform such checks, i.e., it wouldn't be possible to do these checks when adding the `virtual-security-domain` resource (because the corresponding `trusted-virtual-security-domains` wouldn't have been configured yet) or when adding a `security-domain` resource. This would need to be something that's analyzed after the whole subsystem has been configured.